### PR TITLE
conditionals

### PR DESCRIPTION
**Pull request by Pavel-Durov**
-  if ... else statements
- `eq` op
- `neq` op
- function lang tests

### Conditionals

Conditional statements allow a block of code to be executed if a condition is true or if specified, execution of an alternative block. 

#### if...else statements

```
if (true) {
  /* code to run if condition is true */
} else {
  /* run some other code instead */
}
```

`else` block is optional, one can use if statements without it. 

Example:

```
if (true) {
  /* code to run if condition is true */
}
```

#### New Operators
| ==   | Equal | `1 == 2`|
| !=   | Not Equal | `2 != 1`|
| \|\|   | Or | `true \|\| false`|
| &&   | And | `true && false`|

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -78,12 +78,21 @@ Example:
...
 | >   | Greater than | `1 > 2`|
-| <   |  Less than  a| `2 < 1`|
+| <   |  Less than | `2 < 1`|
+
+#### Logical Operators
+
+| Symbol |  Meaning  | Example |
+|:-----|:--------:|------:|
+| ==   | Equal | `1 == 2`|
+| !=   | Not Equal | `2 < 1`|
```
Comment (vext01) [inline README.md:93]: Example uses the wrong operator.

File: README.md [hunk 2]
```diff
@@ -146,6 +155,30 @@ let x = 2;
 let y = 1 * _x;
 ```
 
+### Conditionals
+
+Conditional statements allow a block of code to be executed if a condition is true or if specified, execution of an alternative block.
```
Comment (vext01) [inline README.md:160]: This sentence could be better worded I think.

File: src/instruction.rs [hunk 3]
```diff
@@ -24,6 +25,41 @@ impl StackValue {
             )),
         }
     }
+    pub fn is_same_type(&self, other: &Self) -> bool {
+        match self {
```
Comment (vext01) [inline src/instruction.rs:29]: Could this be shorter by comparing the integer discriminants of the values?

File: src/lib.rs [hunk 4]
```diff
@@ -162,6 +162,78 @@ impl Calc {
...
+                        format!(
+                            "Operand {} and Operand {} cannot be applied to Equal operation",
+                            op1, op2
+                        )
+                        .to_string(),
+                    ));
+                }
+                stack_value
+            }
+            BinaryOp::NotEqual => {
```
Comment (vext01) [inline src/lib.rs:183]: Have you considered implementing `!=` in terms of `==`. Do the same computation, but negative the end result?

**Comment by vext01** [comment]
Mostly LGTM. Just a handful of comments.

File: README.md (hunk 1 above)
Comment (Pavel-Durov) [reply README.md]: Oops. fixed :point_right: 56de71d397e96b4b318e3ca6d79211f9b4d9efa8

File: README.md (hunk 2 above)
Comment (Pavel-Durov) [reply README.md:160]: yeah, you're right.,
I think I will remove it, no other section in my README has this kind of explanation.

File: src/instruction.rs [hunk 5]
```diff
@@ -24,6 +25,41 @@ impl StackValue {
...
+                if let StackValue::Integer(..) = other {
+                    return true;
+                }
+                false
+            }
+        }
+    }
+}
+
+impl PartialEq for StackValue {
```
[the remark this answers is not in the document]
Comment (Pavel-Durov) [reply src/instruction.rs]: removed :point_right: cc23a58255cda6b903d3f3ba64f82b29fb08caea

File: tests/boolean.rs [hunk 6]
```diff
@@ -139,4 +139,123 @@ mod tests {
             _ => panic!("expected bytecodes to be not empty!"),
         }
     }
+
+    #[test]
+    fn bool_eq_bc() {
+        let calc = &mut Calc::new();
+        let ast = calc.from_str("1 == 1;").unwrap();
```
[the remark this answers is not in the document]
Comment (Pavel-Durov) [reply tests/boolean.rs:146]: will be evaluated as expected, added tests and fixed a bug in `neq` :point_right: c70d8a14c01c7ba9e2b8fec7e42c64d4fd6b0543

File: src/lib.rs (hunk 4 above)
Comment (Pavel-Durov) [reply src/lib.rs:170]: Good point. Implemented :point_right: a34a7051380f080122a538d286b5729e1a101de0

File: src/instruction.rs (hunk 3 above)
Comment (Pavel-Durov) [reply src/instruction.rs]: ok, I'll try it :)

Comment (Pavel-Durov) [reply src/instruction.rs]: Looks better! changed :point_right: 3f983651c600b604b3b98a4a36173bd2bbc0a270

**Comment by Pavel-Durov** [comment]
squashed :point_right:  e9b7ca2c988616968ff250e0c359a5e6487a7775

**Comment by Pavel-Durov** [comment]
bors try

**Comment by vext01** [comment]
we can do an r+.

bors try-

**Pull request merged**